graalpython/com.oracle.graal.python/src/com/oracle/graal/python/builtins/objects/foreign/ForeignObjectBuiltins.java

```diff
@@ -1031,6 +1031,28 @@ protected static long doIt(Object object,
                 gil.acquire();
             }
         }
+        @Specialization(limit = "3", guards="lib.fitsInBigInteger(object)")
```

Comment:
If the value does not fit in long nor big integer, then this will cause `UnsupportedSpecializationException`, while previously it would cause Python level exception that the user can handle. Why another `@Specialization` and not another if in the existing specialization?

Reply:
Could we just add a `@Fallback` raising the Python exception?

Reply:
Yes, but I am not sure if this is correct because of what you pointed out in https://github.com/oracle/graalpython/pull/346#discussion_r1279529821. I think that one would just need more complex example than the tests that were added to trigger the potential issue. All in all, adding an `if` to the single `@Specialization` looks like safe and simple solution.

Reply:
OK, merging the two `@Specialization`s back into one: 04bf4c6 - thanks for your review, Štěpáne.

Reply:
Out of curiosity - is there any (expected, of course in practice one would have to measure this) advantage (e.g. in performance) of having 2 separate simpler specializations VS one more complex one?

Reply:
Ah, I forgot about `PythonObjectFactory` that's good point. I'd still keep it this way even though it's a small trade off of readability vs efficiency. Regarding performance: the specialization will get new instance for each interop object "type" and each such instance will specialize to that specific "type", so I don't think that two specializations would gain anything in this case except for the removal of the unnecessary `PythonObjectFactory` allocation. OTOH, more specializations, especially with guards => more complex `execute` and `executeAndSpecialize` => can have impact on the interpreter performance (see also [host inlining](https://www.graalvm.org/latest/graalvm-as-a-platform/language-implementation-framework/HostOptimization/))